python/ray/data/_internal/datasource/parquet_datasink.py

```diff
@@ -75,10 +76,15 @@ def write(
 
         def write_blocks_to_path():
             with self.open_output_stream(write_path) as file:
-                schema = BlockAccessor.for_block(blocks[0]).to_arrow().schema
-                with pq.ParquetWriter(file, schema, **write_kwargs) as writer:
-                    for block in blocks:
-                        table = BlockAccessor.for_block(block).to_arrow()
+                tables = [BlockAccessor.for_block(block).to_arrow() for block in blocks]
+                # We only allow unifying nullable fields by promote_options=`default`
+                unified_schema = pa.unify_schemas(
+                    [table.schema for table in tables], promote_options="default"
+                )
+                with pq.ParquetWriter(file, unified_schema, **write_kwargs) as writer:
+                    for table in tables:
+                        if not table.schema.equals(unified_schema):
+                            table = table.cast(unified_schema)
```

Comment:
nit: I think we can always cast and let PA figure out that it's a no-op when schemas are the same